dcm4che-imageio/src/main/java/org/dcm4che3/imageio/plugins/dcm/DicomImageReader.java

```diff
@@ -181,113 +131,27 @@
-        this.pixelDataVR = src.pixelDataVR;
-        this.pixelDataFile = src.pixelDataFile;
-        this.frames = src.frames;        
-        this.flushedFrames = src.flushedFrames;
-        this.width = src.width;
-        this.height = src.height;
-        this.decompressor = null;
-        this.rle = src.rle;
-        // TODO - examine the patch JpegLS data to see if this still works.
-        this.patchJpegLS = src.patchJpegLS;
-        this.samples = src.samples;        
-        this.banded = src.banded;
-        this.bitsStored = src.bitsStored;
-        this.bitsAllocated = src.bitsAllocated;
-        this.dataType = src.dataType;
-        this.frameLength = src.frameLength;
-        this.pmi = src.pmi;
-    }
-
     @Override
-    public void setInput(Object input, boolean seekForwardOnly,
-            boolean ignoreMetadata) {
+    public void setInput(Object input, boolean seekForwardOnly, boolean ignoreMetadata) {
         super.setInput(input, seekForwardOnly, ignoreMetadata);
         resetInternalState();
-        if (input instanceof InputStream) {
-            try {
-                dis = (input instanceof DicomInputStream)
-                        ? (DicomInputStream) input
-                        : new DicomInputStream((InputStream) input);
-            } catch (IOException e) {
-               throw new IllegalArgumentException(e.getMessage());
-            }
-        } else if (input instanceof DicomMetaData) {
-            DicomMetaData metadata = (DicomMetaData) input;
-            initPixelDataFromAttributes(metadata.getAttributes());
-            initPixelDataFile();
-            setMetadata(metadata);
-        } else {
-            iis = (ImageInputStream) input;
-        }
-    }
 
-    private void initPixelDataFromAttributes(Attributes ds) {
-        VR.Holder holder = new VR.Holder();
-        Object value = ds.getValue(Tag.PixelData, holder);
-        if (value != null) {
-            pixelDataVR = holder.vr;
-            if (value instanceof BulkData) {
-                pixelData = (BulkData) value;
-                pixelDataLength = pixelData.length();
-            } else if( value instanceof byte[] ) {
-                pixeldataBytes = (byte[]) value;
-                pixelDataLength = pixeldataBytes.length;
-            } else { // value instanceof Fragments
-                pixelDataFragments = (Fragments) value;
-                pixelDataLength = -1;
-            }
-        }
-    }
+        this.input = input;
 
-    private void initPixelDataFile() {
-        if (pixelData != null)
-            pixelDataFile = pixelData.getFile();
-        else if (pixelDataFragments != null && pixelDataFragments.size() > 1) {
-            Object frag = pixelDataFragments.get(1);
-            if( frag instanceof BulkData ) {
-                pixelDataFile = ((BulkData) frag).getFile();
-            }
+        if (input instanceof DicomMetaData) {
+            setMetadata((DicomMetaData) input);
```

Comment:
I like this setup to handle input being a dicom meta data - that feels like a reasonable way to work with this.

Reply:
Thanks.  A lot of my changes are around making DicomImageReader more flexible, and providing client applications more options in using it.